cmd/kubeadm/app/master/addons.go

```diff
@@ -0,0 +1,267 @@
+							Path:   "/readiness",
+							Port:   intstr.FromInt(8081),
+							Scheme: api.URISchemeHTTP,
+						},
+					},
+					InitialDelaySeconds: 30,
+					TimeoutSeconds:      5,
+				},
+				Ports: []api.ContainerPort{
+					{
+						ContainerPort: kubeDNSPort,
+						Name:          "dns-local",
+						Protocol:      api.ProtocolUDP,
+					},
+					{
+						ContainerPort: kubeDNSPort,
+						Name:          "dns-tcp-local",
+						Protocol:      api.ProtocolTCP,
+					},
+				},
+			},
+			// dnsmasq
+			{
+				Name:  "dnsmasq",
+				Image: images.GetAddonImage(images.KubeDNSmasqImage),
+				Resources: api.ResourceRequirements{
+					Limits:   dnsPodResources,
+					Requests: dnsPodResources,
+				},
+				Args: []string{
+					"--cache-size=1000",
+					"--no-resolv",
+					fmt.Sprintf("--server=127.0.0.1#%d", kubeDNSPort),
+				},
+				Ports: []api.ContainerPort{
+					{
+						ContainerPort: dnsmasqPort,
+						Name:          "dns",
+						Protocol:      api.ProtocolUDP,
+					},
+					{
+						ContainerPort: dnsmasqPort,
+						Name:          "dns-tcp",
+						Protocol:      api.ProtocolTCP,
+					},
+				},
+			},
+			// healthz
+			{
+				Name:  "healthz",
+				Image: images.GetAddonImage(images.KubeExechealthzImage),
+				Resources: api.ResourceRequirements{
+					Limits:   healthzPodResources,
+					Requests: healthzPodResources,
+				},
+				Args: []string{
+					nslookup,
+					"-port=8080",
+					"-quiet",
+				},
+				Ports: []api.ContainerPort{{
+					ContainerPort: 8080,
+					Protocol:      api.ProtocolTCP,
+				}},
+			},
+		},
+		DNSPolicy: api.DNSDefault,
+	}
+
+}
+
+func createKubeDNSServiceSpec(s *kubeadmapi.KubeadmConfig) (*api.ServiceSpec, error) {
+	ip, err := ipallocator.GetIndexedIP(&s.InitFlags.Services.CIDR, 10)
+	if err != nil {
+		return nil, fmt.Errorf("unable to allocate IP address for kube-dns addon from the given CIDR (%q) [%s]", s.InitFlags.Services.CIDR, err)
+	}
+
+	svc := &api.ServiceSpec{
+		Selector: map[string]string{"name": "kube-dns"},
+		Ports: []api.ServicePort{
+			{Name: "dns", Port: 53, Protocol: api.ProtocolUDP},
+			{Name: "dns-tcp", Port: 53, Protocol: api.ProtocolTCP},
+		},
+		ClusterIP: ip.String(),
+	}
+
+	return svc, nil
+}
+
+func CreateEssentialAddons(s *kubeadmapi.KubeadmConfig, client *clientset.Clientset) error {
+	arches := [3]string{"amd64", "arm", "arm64"}
+
+	for _, arch := range arches {
+		kubeProxyDaemonSet := NewDaemonSet(kubeProxy+"-"+arch, createKubeProxyPodSpec(s, arch))
+		SetMasterTaintTolerations(&kubeProxyDaemonSet.Spec.Template.ObjectMeta)
+
+		if _, err := client.Extensions().DaemonSets(api.NamespaceSystem).Create(kubeProxyDaemonSet); err != nil {
+			return fmt.Errorf("<master/addons> failed creating essential kube-proxy addon [%s]", err)
+		}
+	}
+
+	fmt.Println("<master/addons> created essential addon: kube-proxy")
+
+	kubeDNSDeployment := NewDeployment("kube-dns", 1, createKubeDNSPodSpec(s))
+	SetMasterTaintTolerations(&kubeDNSDeployment.Spec.Template.ObjectMeta)
```

Comment:
For efficiency sake, it might be best to pin it to the master.  If it gets scheduled on another node instead of the master then it is taking up space that could be used by a workload.